src/replay/consensus.zig

```diff
@@ -0,0 +1,1321 @@
+/// - Updates fork choice with the new root.
+///
+/// Analogous to [check_and_handle_new_root](https://github.com/anza-xyz/agave/blob/ccdcdbe9b6ff7dbd583d2101fe57b7cc41a6f863/core/src/replay_stage.rs#L4002)
+fn checkAndHandleNewRoot(
+    allocator: std.mem.Allocator,
+    blockstore_reader: *BlockstoreReader,
+    slot_tracker: *SlotTracker,
+    progress: *ProgressMap,
+    fork_choice: *ForkChoice,
+    new_root: Slot,
+) !void {
+    // get the root bank before squash.
+    if (slot_tracker.slots.count() == 0) return error.EmptySlotTracker;
+    var root_tracker = slot_tracker.slots.get(new_root) orelse return error.MissingSlot;
+    const maybe_root_hash, var hash_lg = root_tracker.state.hash.readWithLock();
+    defer hash_lg.unlock();
+    const root_hash = maybe_root_hash.* orelse return error.MissingHash;
+
+    const rooted_slots = try slot_tracker.parents(allocator, new_root);
+    defer allocator.free(rooted_slots);
+
+    // TODO implement leader_schedule_cache.set_root.
+    // TODO have this a seperate function?
+    {
+        var lowest_cleanup_slot = RwMux(Slot).init(0);
+        var max_root = std.atomic.Value(Slot).init(0);
+        var registry = sig.prometheus.Registry(.{}).init(allocator);
+        defer registry.deinit();
+
+        var writer = LedgerResultWriter{
+            .allocator = allocator,
+            .db = blockstore_reader.db,
+            .logger = .noop,
+            .lowest_cleanup_slot = &lowest_cleanup_slot,
+            .max_root = &max_root,
+            .scan_and_fix_roots_metrics = try registry.initStruct(
+                sig.ledger.result_writer.ScanAndFixRootsMetrics,
+            ),
+        };
+
+        try writer.setRoots(rooted_slots);
+    }
+
+    // Audit: The rest of the code maps to Self::handle_new_root in Agave.
+    // TODO
+    // - Prune program cache bank_forks.read().unwrap().prune_program_cache(new_root);
+    // - Set root on slot tracker: let removed_banks = bank_forks.write().unwrap().set_root(
+    //        new_root,
+    //        snapshot_controller,
+    //        highest_super_majority_root,
+    //    )?
+    //
+    // Update the progress map.
+    // TODO Move to its own function?
+    {
+        var to_remove = try std.ArrayListUnmanaged(Slot).initCapacity(
+            allocator,
+            progress.map.count(),
+        );
+        defer to_remove.deinit(allocator);
+
+        var it = progress.map.iterator();
+        while (it.next()) |entry| {
+            // TODO should frozen state be taking into consideration.
+            if (slot_tracker.slots.get(entry.key_ptr.*) == null) {
+                to_remove.appendAssumeCapacity(entry.key_ptr.*);
+            }
+        }
+
+        for (to_remove.items) |key| {
+            _ = progress.map.swapRemove(key);
+        }
+    }
+
+    // Update forkchoice
+    try fork_choice.setTreeRoot(&.{
+        .slot = new_root,
+        .hash = root_hash,
+    });
+}
+
+fn resetFork(
```

Comment:
Can you add a comment saying this is a placeholder for logic that is necessary for certain circumstances we're not currently handling, and mention those circumstances (voting and producing blocks?), and link to the analogous code in agave.

Reply:
[bfdf54e](https://github.com/Syndica/sig/pull/762/commits/bfdf54e2082c6d6319419e5b31a7e7eedda64a23)